grease-exe/src/Grease/Main.hs

```diff
@@ -195,22 +200,43 @@ import What4.ProgramLoc qualified as W4
 import What4.Protocol.Online qualified as W4
 import Prelude (Int, Integer, Integral, Num (..), fromIntegral)
 
+{- Note [Explicitly listed errors]
+
+In this module, we frequently explicitly list all constructors of an error type,
+even when performing the same action for many or all of them. This is so that
+we get a compile error when new error constructors are added and can make an
+informed judgment about whether they are in fact user errors, internal GREASE
+errors, malformed inputs, etc.
+
+-}
+
 -- | Results of analysis, one per given 'Entrypoint'
 newtype Results = Results {getResults :: Map Entrypoint Batch}
 
 doLog :: MonadIO m => GreaseLogAction -> Diag.Diagnostic -> m ()
 doLog la diag = LJ.writeLog la (MainDiagnostic diag)
 
+malformedElf :: GreaseLogAction -> PP.Doc Void -> IO a
+malformedElf la doc = do
+  doLog la (Diag.MalformedElf doc)
+  throw (GreaseException (Text.pack (show doc)))
+
+userError :: GreaseLogAction -> PP.Doc Void -> IO a
+userError la doc = do
+  doLog la (Diag.UserError doc)
```

Comment:
sorry is it normal to log an exception then throw an exception with the same text, feels a bit weird since theyll both land in stderr?

Reply:
Good catch. I was planning on reworking this later, but I might as well do so now. The ideal outcome here IMO is to simply `exit(1)` here instead of throwing an exception. The problem is that the test suite doesn't capture logs that happen before exceptions, which we would want to match on. That's fixable though, and worth fixing.